Track non-technical macro translation skip reasons

SilverMight opened the pull request:
This change modifies `predicates/interface_equivalent.py` to provide specific feedback on why a macro is not translatable. It returns a reason from the `IEResult` enum (or `IEResult.VALID` if the macro is translatable).

The `MacroTranslator` class now has the responsibility of filtering out macros (originally done in `ie_pd`) so it can record any skip reason in the output CSV and the frequency count outputted at the end of translation.

Note that currently, only the first condition that makes a macro untranslatable is reported. We loop through every invocation, and every possible condition, and return the first failure if any.

SilverMight committed:
Track non-technical limitation macro translation skip reasons

Modify the interface equivalent predicate to return an IEResult
detailing why a macro is not translatable, otherwise return none instead
of returning True/False.

Hand off the TLNA macros instead of the already interface-equivalent
filtered macros to MacroTranslator, and have get_macro_record return a
SkipReason with a new optional IEResult if the macro could not be
translated.

Output in the total count at the end of the program and put the IEResult
in the CSV if there is one instead of a SkipReason

SilverMight committed:
Modify predicate names

PappasBrent commented on analyze_transformations.py:
```diff
@@ -114,11 +113,5 @@ def get_interface_equivalent_preprocessordata(results_file: str) -> Preprocessor
 
     # ie_pd only records preprocessor data about interface-equivalent
     # macros
```
```suggestion
```

PappasBrent commented on predicates/interface_equivalent.py:
```diff
@@ -1,78 +1,66 @@
-            all([
-                # Valid for analysis
-                i.HasSemanticData,
+                (not i.MustCreateThunksToTransform, IEResult.THUNKIZING),
```
On Monday let's talk about the predicates; the ones that we use in the MerC formalism are slightly different from the ones in the Maki paper and so we probably won't want to use the same groupings (e.g., thunkizing macros would fall under the "called by name" category in the MerC formalism). Let's discuss tomorrow in person so we can share our thoughts more quickly.

SilverMight replied:
I meant to shoot you a text saying I wanted to talk about some of the predicates - yeah discussing in person would be a good idea.

PappasBrent reviewed (commented):
Small change to remove a comment and then let's talk about the properties tomorrow

SilverMight committed:
Remove outdated comment

Co-authored-by: Brent Pappas <pappasbrent@gmail.com>

SilverMight committed:
Consolidate conditions and IEResult reasons.

Create new properties on macros to match formalism by combining existing
properties (i.e IsCalledByName), and update enum to have matching reason.
Seperate out macros being called where modifiable / addressable values
are required.
Move IsExpansionControlFlowStmt to MustUseMetaprogrammingToTransform,
and add IsNamePresentInCPPConditional as well

SilverMight committed:
Rename has semantic data to syntactically invalid property
Also add TODO to implement compound statement handling

SilverMight committed:
Update MerC to match formalize and centralize translation logic

Handle all translation logic in interface_equivalent.py, instead of
doing checks in MacroTranslator. The only checks remaining in macro
translator are technical skip reasons which are now seperated

ie_def now returns an IEResult AND TranslationTarget. If
IEResult is Valid, the macro satisifies all properties to be translated
to the target.

Split up previously grouped reasonings (i.e CANNOT TRANSFORM) into many reasons to match the
formalism in the research paper.

Translation logic changes:
* Check that all invocations are not expanded where constexpr required
  (excluding enums)
* Check that arguments aren't expanded where constexpr required
* Check on expansion types being void (fixed some incorrect
  translations that compiled, i.e UNUSED macros)
* Seperate argument checks and void/non-void function checks

Other changes:
* Remove unused functions in analysis (easy to transform def/invocation)
* CanBeTurnedIntoVariable (now unused by MerC translation logic) does
  not check arguments using DeclRefExpr (unnecessary, and also okay for
  functions to have)

SilverMight committed:
Add check to ensure body doesn't end with compound stmt.

Also, remove unused IEResults

SilverMight committed:
Simplify SkipRecord

SilverMight committed:
Optimize imports

PappasBrent reviewed (approved):
LGTM, thanks for the careful work matching the implementation to the formalism.

PappasBrent commented on translationstats.py:
```diff
@@ -13,10 +13,7 @@ class MacroType(Enum):
-
-class SkipType(TranslatorAction):
+class TechnicalSkip(Enum):
     NOT_INTERFACE_EQUIVALENT = auto()
```
I think you can remove this skip record type since it seems unused.

SilverMight replied:
It's still used in `MacroTranslator::should_skip_due_to_technical_limitations`

SilverMight replied:
Oh my bad - you mean `NOT_INTERFACE_EQUIVALENT`, good catch

SilverMight committed:
Remove old debug print statements

SilverMight committed:
Ensure that variables expansion type is not void

SilverMight committed:
Remove unused technical skip type
